datafusion/core/src/datasource/flight/mod.rs

```diff
@@ -0,0 +1,259 @@
+}
+
+impl FlightTableFactory {
+    /// Create a data source using the provided driver
+    pub fn new(driver: Arc<dyn FlightDriver>) -> Self {
+        Self { driver }
+    }
+
+    /// Convenient way to create a [FlightTable] programatically, as an alternative to DDL.
+    pub async fn open_table(
+        &self,
+        entry_point: impl Into<String>,
+        options: HashMap<String, String>,
+    ) -> datafusion_common::Result<FlightTable> {
+        let origin = entry_point.into();
+        let channel = Channel::from_shared(origin.clone())
+            .unwrap()
+            .connect()
+            .await
+            .map_err(|e| DataFusionError::External(Box::new(e)))?;
+        let metadata = self
+            .driver
+            .metadata(channel.clone(), &options)
+            .await
+            .map_err(|e| DataFusionError::External(Box::new(e)))?;
+        let logical_schema = metadata.plan_properties.eq_properties.schema().clone();
+        Ok(FlightTable {
+            driver: self.driver.clone(),
+            channel,
+            options,
+            origin,
+            logical_schema,
+        })
+    }
+}
+
+#[async_trait]
+impl TableProviderFactory for FlightTableFactory {
+    async fn create(
+        &self,
+        _state: &dyn Session,
+        cmd: &CreateExternalTable,
+    ) -> datafusion_common::Result<Arc<dyn TableProvider>> {
+        let table = self.open_table(&cmd.location, cmd.options.clone()).await?;
+        Ok(Arc::new(table))
+    }
+}
+
+/// Extension point for integrating any Flight RPC service as a [FlightTableFactory].
+/// Handles the initial `GetFlightInfo` call and all its prerequisites (such as `Handshake`),
+/// to produce a [FlightMetadata].
+#[async_trait]
+pub trait FlightDriver: Sync + Send + Debug {
+    /// Returns a [FlightMetadata] from the specified channel,
+    /// according to the provided table options.
+    /// The driver must provide at least a [FlightInfo] in order to construct a flight metadata.
+    async fn metadata(
+        &self,
+        channel: Channel,
+        options: &HashMap<String, String>,
+    ) -> arrow_flight::error::Result<FlightMetadata>;
+}
+
+/// The information that a [FlightDriver] must produce
+/// in order to register flights as DataFusion tables.
+#[derive(Clone, Debug)]
+pub struct FlightMetadata {
+    /// FlightInfo object produced by the driver
+    pub(super) flight_info: Arc<FlightInfo>,
+    /// Physical plan properties. Sensible defaults will be used if the
+    /// driver doesn't need (or care) to customize the execution plan.
+    pub(super) plan_properties: Arc<PlanProperties>,
+    /// The gRPC headers to use on the `DoGet` calls
+    pub(super) grpc_metadata: Arc<MetadataMap>,
+}
+
+impl FlightMetadata {
+    /// Provide custom [PlanProperties] to account for service specifics,
+    /// such as known partitioning scheme, unbounded execution mode etc.
+    pub fn new(info: FlightInfo, props: PlanProperties, grpc: MetadataMap) -> Self {
```

Comment:
this is really cool -- we actually implemented some version of this custom for InfluxDB 3.0 for services that exchange data between themselves